server/src/main/java/org/apache/iotdb/db/engine/compaction/TsFileManagement.java

```diff
@@ -66,325 +62,464 @@
+    logger.debug("{} [compaction] merge starts to delete file list", storageGroupName);
+    if (sequence) {
+      if (sequenceTsFileResources.containsKey(timePartitionId)) {
+        if (sequenceTsFileResources.get(timePartitionId).size() > level) {
+          sequenceTsFileResources.get(timePartitionId).get(level).removeAll(mergeTsFiles);
         }
-        try {
-          seqFile.getModFile().close();
-        } catch (IOException e) {
-          logger.error(
-              "Cannot close the ModificationFile {}", seqFile.getModFile().getFilePath(), e);
+      }
+    } else {
+      if (unSequenceTsFileResources.containsKey(timePartitionId)) {
+        if (unSequenceTsFileResources.get(timePartitionId).size() > level) {
+          unSequenceTsFileResources.get(timePartitionId).get(level).removeAll(mergeTsFiles);
         }
       }
-    } catch (IOException e) {
-      logger.error(
-          "{} cannot clean the ModificationFile of {} after merge",
-          storageGroupName,
-          seqFile.getTsFile(),
-          e);
     }
   }
 
-  private void removeMergingModification() {
+  private void deleteLevelFile(TsFileResource seqFile) {
+    seqFile.writeLock();
     try {
-      if (mergingModification != null) {
-        mergingModification.remove();
-        mergingModification = null;
-      }
+      ChunkCache.getInstance().clear();
+      TimeSeriesMetadataCache.getInstance().clear();
+      FileReaderManager.getInstance().closeFileAndRemoveReader(seqFile.getTsFilePath());
+      seqFile.setDeleted(true);
+      seqFile.delete();
     } catch (IOException e) {
-      logger.error("{} cannot remove merging modification ", storageGroupName, e);
+      logger.error(e.getMessage(), e);
+    } finally {
+      seqFile.writeUnlock();
     }
   }
 
-  public void mergeEndAction(
-      List<TsFileResource> seqFiles, List<TsFileResource> unseqFiles, File mergeLog) {
-    logger.info("{} a merge task is ending...", storageGroupName);
+  public static int getMergeLevel(File file) {
+    String mergeLevelStr =
+        file.getPath()
+            .substring(file.getPath().lastIndexOf(FILE_NAME_SEPARATOR) + 1)
+            .replaceAll(TSFILE_SUFFIX, "");
+    return Integer.parseInt(mergeLevelStr);
+  }
 
-    if (Thread.currentThread().isInterrupted() || unseqFiles.isEmpty()) {
-      // merge task abort, or merge runtime exception arose, just end this merge
-      logger.info("{} a merge task abnormally ends", storageGroupName);
-      return;
+  private List<SortedSet<TsFileResource>> newSequenceTsFileResources(Long k) {
+    List<SortedSet<TsFileResource>> newSequenceTsFileResources = new ArrayList<>();
+    for (int i = 0; i < seqLevelNum; i++) {
+      newSequenceTsFileResources.add(
+          new TreeSet<>(
+              (o1, o2) -> {
+                try {
+                  int rangeCompare =
+                      Long.compare(
+                          Long.parseLong(o1.getTsFile().getParentFile().getName()),
```

Comment:
I doubt the rangeCompare of the parentFile，as the current file dir is like sgName/timepartition/virtualSG